### PR TITLE
Bug 2093462: status: Watch clusteroperators

### PR DESCRIPTION
**Pull request by Miciah**
Add a watch on clusteroperators in the status controller so that it updates or recreates the ingress clusteroperator as necessary should something else update or delete it.

* `manifests/00-cluster-role.yaml`: Allow the operator to list and watch clusteroperators.
* `pkg/operator/controller/status/controller.go` (`New`): Watch clusteroperators with a map function and predicate to reconcile the default ingresscontroller if `clusteroperators.config.openshift.io/ingress` is updated.
* `test/e2e/all_test.go` `(TestAll`): Add the new `TestOperatorRecreatesItsClusterOperator` test to the serial tests.
* `test/e2e/operator_test.go` (`TestOperatorRecreatesItsClusterOperator`): New test.  Verify that the operator recreates the "ingress" clusteroperator if the clusteroperator is deleted.

---

This PR is the cluster-ingress-operator equivalent of <https://github.com/openshift/cluster-dns-operator/pull/261>.

This PR supersedes <https://github.com/openshift/cluster-ingress-operator/pull/702>.  Cc: @mfojtik

### CODE REVIEW

**Comment by gcs278** [comment]
This will need to be added to `all_tests.go` due to E2E updates

**Comment by Miciah** [comment]
Updated for #756.

**Comment by Miciah** [comment]
```
         --- FAIL: TestAll/parallel/TestHAProxyTimeouts (35.86s)
        --- FAIL: TestAll/parallel/TestHAProxyTimeoutsRejection (38.09s) 
```
/test e2e-aws-operator

**Comment by Miciah** [comment]
Must-gather.
/test e2e-aws-operator

```
Failing invariants:

[bz-monitoring][invariant] alert/Watchdog must have no gaps or changes
```
/test e2e-aws-single-node

**Comment by candita** [comment]
I think the bindata.go update was missing from https://github.com/openshift/cluster-dns-operator/pull/261

File: pkg/operator/controller/status/controller.go [hunk 1]
```diff
@@ -65,9 +68,32 @@ func New(mgr manager.Manager, config Config) (controller.Controller, error) {
...
+				Namespace: config.Namespace,
+				Name:      manifests.DefaultIngressControllerName,
+			},
+		}}
+	}
+	if err := c.Watch(
+		&source.Kind{Type: &configv1.ClusterOperator{}},
+		// The status controller doesn't care which ingresscontroller it
+		// is reconciling, so just enqueue a request to reconcile the
+		// default ingresscontroller.
```
[the remark this answers is not in the document]
Comment (Miciah) [reply pkg/operator/controller/status/controller.go:91]: Right, the status controller just needs some ingresscontroller for its reconciliation request, which will cause it to update the "ingress" clusteroperator, and there should always be a "default" ingresscontroller.

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -299,6 +299,34 @@ func TestCustomIngressClass(t *testing.T) {
...
+// automatically (which means that other tests could interfere with this one).
+func TestOperatorRecreatesItsClusterOperator(t *testing.T) {
+	co := &configv1.ClusterOperator{}
+	coName := controller.IngressClusterOperatorName()
+	if err := kclient.Get(context.TODO(), coName, co); err != nil {
+		t.Fatalf("failed to get clusteroperator %q: %v", coName.Name, err)
+	}
+	if err := kclient.Delete(context.TODO(), co); err != nil {
+		t.Fatalf("failed to delete clusteroperator %q: %v", coName.Name, err)
+	}
```
[the remark this answers is not in the document]
Comment (Miciah) [reply test/e2e/operator_test.go:317]: I could add a check to verify that the UID changed.

**Comment by Miciah** [comment]
> I think the bindata.go update was missing from [openshift/cluster-dns-operator#261](https://github.com/openshift/cluster-dns-operator/pull/261)

The ingress operator only includes `manifests/` in its `bindata.go` because the `ingress-operator` executable includes a subcommand that emits the manifests, which the installer uses to provide manifests that the user can edit to customize them at install-time.  See <https://github.com/openshift/cluster-ingress-operator/pull/309/commits/c77d6aa8e8356d10fbc0ae662ba03a0bbb505a4f>.

**Comment by Miciah** [comment]
Last push updates `TestOperatorRecreatesItsClusterOperator` to check the clusteroperator's deletion timestamp and uid to verify that it really has been recreated.

**Comment by Miciah** [comment]
Rebased.

File: pkg/operator/controller/status/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (candita) [reply pkg/operator/controller/status/controller.go:91]: If we accept https://github.com/openshift/cluster-ingress-operator/pull/778, there might not be a default ingresscontroller, but I think more work probably needs to be done on that.

**Pull request merged**